### ISSUE TITLE
Incorrect IMC parameter detection on SkyLake and KabyLake CPUs

### ISSUE THREAD

**Issue by starslab** [memtest86plus/memtest86plus#375]
I've got two Gigabyte boards, z270 with i5-7600, and z170 with i7-6700k.

On both these boards memtest86+ 7.0.0 reports different IMC clock speeds and DDR4-datarate than the UEFI and memtest86.

Is there any way I can gather data to help with correcting this?

**Comment by starslab** [comment]
Some speeds are detected correctly, like DDR4-2133.

I don't know if this is within the scope of what you're wanting to detect/display, but this CPU is underclocked and has HyperThreading disabled, and that's not reflected correctly here either.


![IMG_20240117_210700](https://github.com/memtest86plus/memtest86plus/assets/6814571/dd4756fe-ef87-4416-9884-6680eb46750f)

[cpuz.zip](https://github.com/memtest86plus/memtest86plus/files/13972401/cpuz.zip)

**Comment by starslab** [comment]
DDR4-2200 is shown incorrectly

![IMG_20240117_210249](https://github.com/memtest86plus/memtest86plus/assets/6814571/0c627670-f3fc-4f7f-aab1-41a9be0c9693)

[cpuz.zip](https://github.com/memtest86plus/memtest86plus/files/13972419/cpuz.zip)

**Comment by starslab** [comment]
As is DDR4-2266

![IMG_20240117_201627](https://github.com/memtest86plus/memtest86plus/assets/6814571/93c5be1a-65f0-4193-9794-07ea8a4d1746)

[cpuz.zip](https://github.com/memtest86plus/memtest86plus/files/13972428/cpuz.zip)

**Comment by starslab** [comment]
And DDR4-2300

![IMG_20240117_200535](https://github.com/memtest86plus/memtest86plus/assets/6814571/455f9502-9184-4941-914a-cfb44e91bfb6)

[cpuz.zip](https://github.com/memtest86plus/memtest86plus/files/13972431/cpuz.zip)

**Comment by starslab** [comment]
DDR4-2400 shows correctly.

![IMG_20240117_211135](https://github.com/memtest86plus/memtest86plus/assets/6814571/071c66df-2b17-4cc4-ac76-54929ebb634f)

[cpuz.zip](https://github.com/memtest86plus/memtest86plus/files/13972434/cpuz.zip)

**Comment by starslab** [comment]
I can get more if that would be of value.

**Comment by x86fr** [comment]
Thanks for these reports. The issue seems to be with the uncommon 100:100 ratio instead of the standard 100:133 ratio.

I'll try to correct this soon.

**Comment by debrouxl** [comment]
> I don't know if this is within the scope of what you're wanting to detect/display, but this CPU is underclocked and has HyperThreading disabled, and that's not reflected correctly here either.

* AFAICS, the fact that HT is disabled is displayed, though maybe not in a very visible way :)
4C/8T is what the processor registers (CPUID), which are unaffected by disabling HT in the BIOS, report; your picture shows that memtest86+ uses 4T, which is undoubtedly what the ACPI MADT produced and exposed by the BIOS contains. The MADT of the Core i7-8750H (6C/12T) laptop I'm using, where I disabled HT in the BIOS as well, contains only 6 Processor Local APIC and 6 Local APIC NMI struct instances, as it should.

* as for the CPU underclocking... nowadays, memtest86+ attempts to measure the CPU clock frequency, using the APIC timer when available (it basically must be on a machine of that generation), otherwise the PIT. Both of those give pretty accurate timings on a crushing majority of computers, and the two reported issues with PIT or CPU speed so far were due to broken firmware (#40) or likely broken hardware (#122). Therefore, on the one hand, I do believe you when you write that you underclocked the CPU, but on the other hand, memtest86+ isn't known to produce improper CPU speed results.
Well, what do the OS and CPU-Z indicate about the CPU speed ? These high-entropy cpuz.cvf binary files probably contain the information, and I hope that an offline CPU-Z can read them... but I don't have a Windows computer for running CPU-Z handy. I'll try to use Wine.

**Comment by starslab** [comment]
> * as for the CPU underclocking... nowadays, memtest86+ attempts to measure the CPU clock frequency, using the APIC timer when available (it basically must be on a machine of that generation), otherwise the PIT. Both of those give pretty accurate timings on a crushing majority of computers, and the two reported issues with PIT or CPU speed so far were due to broken firmware ([PIT Counter broken on Framework Laptop #40](https://github.com/memtest86plus/memtest86plus/issues/40)) or likely broken hardware ([(CPU?) Speed showing incorrectly #122](https://github.com/memtest86plus/memtest86plus/issues/122)). Therefore, on the one hand, I do believe you when you write that you underclocked the CPU, but on the other hand, memtest86+ isn't known to produce improper CPU speed results.
>   Well, what do the OS and CPU-Z indicate about the CPU speed ? These high-entropy cpuz.cvf binary files probably contain the information, but I don't know how to make an offline CPU-Z can read them (if even possible). I don't have a Windows computer for running CPU-Z handy, but for that purpose, it could run on Wine.

![Capture](https://github.com/memtest86plus/memtest86plus/assets/6814571/61a68acd-dc4c-438d-92f1-8168e8a55caf)

**Comment by debrouxl** [comment]
Thanks. Both unsurprising and surprising. Well, memtest86+ has always correctly reported the speed of the dozens of processors I used it with, including the BIOS-downlocked K8 family and BIOS-overclocked Athlon XP processor :)
I'll defer to Sam and Martin.

**Comment by x86fr** [comment]
Measuring clock on very old CPU was extremely easy. It's a now a nightmare when the CPU is under/over-clocked, with dozens of parameters in undocumented MSR/MMR/etc to check. If the clock is not correct, that's because the ACPI correction factor was not applied correctly. But I'm not sure that's the case here. 

Your CPU-Z dump shows that the CPU is limited to 3.4 GHz using Turbo ratios and that the hardware Turbo Management is disabled. Non-Turbo are still at 4 GHz. Windows handle Turbo ratio management, so the CPU is effectively running at 3.4 GHz under Windows. Memtest86+ doesn't handle Turbo Ratio, so it's possible that your CPU is running at it's max non-turbo ratio clock under Memtest86+ (so 4 GHz). 

I'll do some tests to be sure.

**Comment by x86fr** [comment]
Non-Turbo ratio is fuse locked. It can't be changed on production and QS parts (only early ES).

**Comment by Kuntsaito** [comment]
> The issue seems to be with the uncommon 100:100 ratio instead of the standard 100:133 ratio

On the previous generation (Haswell) it's probably the same bug. According to Intel's specifications for entry-level processors (Celeron, Pentium), memory frequency is limited to 1400 MHz.
133 x 12 = 1600
100 x 14 = 1400
133 x 14 = 1866

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/3454497d-5100-4449-b4c6-86d56117d2bc" />

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/fa394cf3-4b79-4d9a-8cf2-8a5de21cbd55" />

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/c86bdba3-299f-4b79-aeb7-f55bba0ef00a" />

**Comment by x86fr** [comment]
That's not very complex to fix. Just need to add the 100/100 vs 133 ratio. I'll try to assemble a HSW and SKL boards with this uncommon ratio asap
